### PR TITLE
Docker App for PHP Test Suite

### PR DESCRIPTION
**Pull request by adrianduffell**
This PR adds a docker app for running the PHP Test Suite.

The configuration files are tucked away under `docker/wc-admin-php-test-suite`. The name is somewhat long but is intended to provide a unique [project name](https://docs.docker.com/compose/#multiple-isolated-environments-on-a-single-host). Inside, a Dockerfile is added to build an image capable of installing the PHP test suite and running PHPUnit. Docker Compose is used to combine it with a MySQL database. A custom `ENTRYPOINT` script is used to perform installation on first-use.

For convenience, an NPM script is provided to run the tests:

```
npm run test:php
```

This will run the installation on first-use and perform testing with PHPUnit. Subsequent runs will skip installation.

A complexity arises where re-installation is required to update the WordPress or WooCommerce dependencies to the latest version. For now, re-installation can be triggered by removing the `test-suite` volume with Docker. Instructions for this are provided in the README. As a next step, it might be worth making an npm script like `test:php:re-install` to make this more convenient.

A minor change is made to the installation script to allow the WooCommerce plugin location to be specified from an environment variable.

The [`FS_METHOD` constant](https://wordpress.org/support/article/editing-wp-config-php/#wordpress-upgrade-constants) is added to the test bootstrap. This ensures the theme update tests modify files directly. Without it, WordPress is unable to determine the correct method and results in tests failing. I think this is because the container has files owned by root due to default Docker behavior.

### CODE REVIEW

**Comment by adrianduffell** [comment]
Some tests fail which I plan to investigate further.

```
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.3.23
Configuration: /plugin/phpunit.xml.dist

..................................F.F........F.F.FF............  63 / 302 ( 20%)
............................................................... 126 / 302 ( 41%)
............................................................... 189 / 302 ( 62%)
..........................F.F.................................. 252 / 302 ( 83%)
..................................................              302 / 302 (100%)
```

File: docker/phpunit/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	php7-pdo_mysql \
+	php7-phar \
+	php7-session \
+	php7-soap \
+	php7-tokenizer \
+	php7-xdebug \
+	php7-xml \
+	php7-xmlreader \
+	php7-zip \
+	php7-zlib
```
Comment (samueljseay) [inline docker/phpunit/Dockerfile:42]: Love it :D 

I like the choice of Alpine and 👏 for figuring out the extensive list of needed dependencies. 🤯

File: docker-compose-phpunit.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+version: "3"
```
Comment (samueljseay) [inline docker-compose-phpunit.yml:1]: I wonder if we should tuck files like this away under the docker sub directory just to keep from cluttering the root dir?

File: package.json [hunk 3]
```diff
@@ -71,7 +71,10 @@
 		"storybook": "start-storybook  -c ./storybook -p 6007 --ci",
 		"build-storybook": "build-storybook  -c ./storybook",
 		"changelog": "node ./bin/changelog --changelogSrcType='ZENHUB_RELEASE'",
-		"bump-version": "npm run -s install-if-deps-outdated && php ./bin/update-version.php"
+		"bump-version": "npm run -s install-if-deps-outdated && php ./bin/update-version.php",
+		"predocker-phpunit:install": "docker-compose -f docker-compose-phpunit.yml down && docker volume rm -f woocommerce-admin_testsuite",
```
Comment (samueljseay) [inline package.json:75]: I wonder if its fast enough to do this that we could just wrap up the running of all these scripts into a single bash script or something? just to keep the cognitive overhead of running these tests in docker low?

Comment (samueljseay) [reply package.json:75]: It'd be so cool if non docker uses tried running a single script like this and the speed/convenience got some converts to docker 😂

Comment (adrianduffell) [reply package.json:75]: Nice idea!

It takes a few minutes on my machine to complete the install script. However I think in a single script, we could detect if installation has been run- and if not then automatically start the installation before it proceeds with the tests 🤔

**Comment by senadir** [comment]
Throwing this here as an example of how we do e2e and phpunit using Docker, locally and in Travis.
https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2004/files

**Comment by samueljseay** [comment]
@adrianduffell When I run this locally, it does not show the test output, it seems the container just completes any output. Maybe we're not keeping the container alive properly?

**Comment by adrianduffell** [comment]
> @adrianduffell When I run this locally, it does not show the test output, it seems the container just completes any output. Maybe we're not keeping the container alive properly?

I think there is an issue in the test scripts preventing PHPUnit running locally. It affects VVV as well. I have a fix in PR https://github.com/woocommerce/woocommerce-admin/pull/5650.

File: package.json (hunk 3 above)
Comment (adrianduffell) [reply package.json:75]: @samueljseay What do you think about addressing this in a follow-up PR in the next cooldown instead?

File: docker-compose-phpunit.yml (hunk 2 above)
Comment (adrianduffell) [reply docker-compose-phpunit.yml]: Sounds good, now done in 7d912f7

File: package.json (hunk 3 above)
Comment (samueljseay) [reply package.json]: Yep sounds good 👍

**Comment by adrianduffell** [comment]
I refreshed this a little today.

I'm now getting output about the tests running. I still need to investigate the failing ones.

```
$ npm run docker-phpunit:test

> @woocommerce/admin-library@1.7.0-dev docker-phpunit:test /Users/adrian/vagrant-local/www/wordpress-one/public_html/wp-content/plugins/woocommerce-admin
> docker-compose -p woocommerce-admin -f docker/docker-compose-phpunit.yml run --rm phpunit

Creating woocommerce-admin_phpunit_run ... done
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
WordPress database error Table 'wordpress_test.wptests_wc_admin_notes' doesn't exist for query SELECT note_id FROM wptests_wc_admin_notes WHERE name = 'wc-admin-deactivate-plugin' ORDER BY note_id ASC made by PHPUnit\TextUI\Command::main, PHPUnit\TextUI\Command->run, PHPUnit\TextUI\Command->handleArguments, PHPUnit\TextUI\Command->handleBootstrap, PHPUnit\Util\FileLoader::checkAndLoad, PHPUnit\Util\FileLoader::load, include_once('/plugin/tests/bootstrap.php'), WC_Admin_Unit_Tests_Bootstrap::instance, WC_Admin_Unit_Tests_Bootstrap->__construct, require_once('/tmp/wordpress-tests-lib/includes/bootstrap.php'), require_once('wp-settings.php'), do_action('plugins_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Automattic\WooCommerce\Packages::on_init, Automattic\WooCommerce\Packages::load_packages, call_user_func, Automattic\WooCommerce\Admin\Composer\Package::init, Automattic\WooCommerce\Admin\Notes\DeactivatePlugin::delete_note, Automattic\WooCommerce\Admin\Notes\Notes::delete_notes_with_name, WC_Data_Store->__call, Automattic\WooCommerce\Admin\Notes\DataStore->get_notes_with_name
<div id="error"><p class="wpdberror"><strong>WordPress database error:</strong> [Table &#039;wordpress_test.wptests_wc_admin_notes&#039; doesn&#039;t exist]<br /><code>SELECT note_id FROM wptests_wc_admin_notes WHERE name = &#039;wc-admin-deactivate-plugin&#039; ORDER BY note_id ASC</code></p></div>Installing WooCommerce and WooCommerce Admin...
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.3.23
Configuration: /plugin/phpunit.xml.dist

..................................F.F........F.F.FF............  63 / 307 ( 20%)
............................................................... 126 / 307 ( 41%)
............................................................... 189 / 307 ( 61%)
.F...........................F.F............................... 252 / 307 ( 82%)
.......................................................         307 / 307 (100%)

Time: 5.61 minutes, Memory: 101.00 MB

There were 9 failures:

1) WC_Tests_API_Onboarding_Themes::test_install_theme
Failed asserting that 500 matches expected 200.

/plugin/tests/api/onboarding-themes.php:60

2) WC_Tests_API_Onboarding_Themes::test_activate_theme
Failed asserting that 500 matches expected 200.

/plugin/tests/api/onboarding-themes.php:101

3) WC_Tests_API_Plugins::test_install_plugin
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => 'facebook-for-woocommerce'
 )

/plugin/tests/api/plugins.php:60

4) WC_Tests_API_Plugins::test_activate_plugin
Failed asserting that an array contains 'facebook-for-woocommerce'.

/plugin/tests/api/plugins.php:95

5) WC_Tests_API_Plugins::test_install_non_allowed_plugins
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => 'facebook-for-woocommerce'
 )

/plugin/tests/api/plugins.php:130

6) WC_Tests_API_Plugins::test_activate_non_allowed_plugins
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => 'facebook-for-woocommerce'
 )

/plugin/tests/api/plugins.php:150

7) WC_Tests_CES_Tracks::test_updating_options_triggers_ces
Failed asserting that a boolean is not empty.

/plugin/tests/features/class-wc-tests-ces-tracks.php:38

8) WC_Admin_Tests_Plugins_Helper::test_get_active_plugin_slugs
Facebook for WooCommerce should be listed as active.
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => 'facebook-for-woocommerce'
 )

/plugin/tests/plugins-helper.php:71

9) WC_Admin_Tests_Plugins_Helper::test_is_plugin_active
Facebook for WooCommerce should be installed.
Failed asserting that false matches expected true.

/plugin/tests/plugins-helper.php:105

FAILURES!
Tests: 307, Assertions: 1720, Failures: 9.
```

**Comment by adrianduffell** [comment]
All but one of the failing tests above are for tests involving theme switching. I'm able to fix it by setting [`FS_METHOD`](https://wordpress.org/support/article/editing-wp-config-php/#wordpress-upgrade-constants) to `direct` in the bootstrap. This might be due to Docker mounting the files as root. I'm going to look into it a bit more before committing this fix.

**Comment by adrianduffell** [comment]
I went ahead with using `FS_METHOD` in e6d8abe87cb7570566c136c94c4223dd70a8ef07. I'm not totally happy about it but thinking it's a good solution for now.

There is now just a single error reported, which only happens on the first run.

```
WordPress database error Table 'wordpress_test.wptests_wc_admin_notes' doesn't exist for query SELECT note_id FROM wptests_wc_admin_notes WHERE name = 'wc-admin-deactivate-plugin' ORDER BY note_id ASC made by PHPUnit\TextUI\Command::main, PHPUnit\TextUI\Command->run, PHPUnit\TextUI\Command->handleArguments, PHPUnit\TextUI\Command->handleBootstrap, PHPUnit\Util\FileLoader::checkAndLoad, PHPUnit\Util\FileLoader::load, include_once('/plugin/tests/bootstrap.php'), WC_Admin_Unit_Tests_Bootstrap::instance, WC_Admin_Unit_Tests_Bootstrap->__construct, require_once('/tmp/wordpress-tests-lib/includes/bootstrap.php'), require_once('wp-settings.php'), do_action('plugins_loaded'), WP_Hook->do_action, WP_Hook->apply_filters, Automattic\WooCommerce\Packages::on_init, Automattic\WooCommerce\Packages::load_packages, call_user_func, Automattic\WooCommerce\Admin\Composer\Package::init, Automattic\WooCommerce\Admin\Notes\DeactivatePlugin::delete_note, Automattic\WooCommerce\Admin\Notes\Notes::delete_notes_with_name, WC_Data_Store->__call, Automattic\WooCommerce\Admin\Notes\DataStore->get_notes_with_name
<div id="error"><p class="wpdberror"><strong>WordPress database error:</strong> [Table &#039;wordpress_test.wptests_wc_admin_notes&#039; doesn&#039;t exist]<br /><code>SELECT note_id FROM wptests_wc_admin_notes WHERE name = &#039;wc-admin-deactivate-plugin&#039; ORDER BY note_id ASC</code></p></div>Installing WooCommerce and WooCommerce Admin...
```

**Comment by adrianduffell** [comment]
The above WordPress table error occurs under VVV too. I've created an [issue](https://github.com/woocommerce/woocommerce-admin/issues/5879) to work on it seperately.

I think this is ready for re-review! 🎉

**Review by louwie17 — APPROVED** [review]
Tested it, and all tests pass, much much faster then my VVV 🚤  🎉  Well done @adrianduffell!

When I first ran `npm run test:php` the mysql didn't quite seem to be booted up yet, so it failed when trying to run the setup script (not a stopper for me).
I fixed it by running `docker volume rm -f wc-admin-php-test-suite_test-suite` after, and then running `npm run test:php`.
We could add something like this postgres example (the wait-for-postgres.sh script): https://docs.docker.com/compose/startup-order/ to the `entrypoint.sh` script.

**Comment by adrianduffell** [comment]
Thanks @louwie17 !

The`wait-for` script sounds good. I tried to solve that problem using the healthcheck command, but seems like it's not quite good enough.

**Pull request merged**